lib/Runner.js

```diff
@@ -71,133 +47,87 @@ function getIndirectDeps(projectRootDir /*: string */) /*: Promise<Object> */ {
   });
 }
 
-function moduleFromFilePath(filePathArg) {
-  var parsed = path.parse(path.normalize(filePathArg));
-  var basename = path.basename(parsed.base, '.elm');
-
-  // Turn these into module name checks to be performed, in order.
-  // e.g. 'tests/All/Passing.elm' ===> ['Passing', 'All', 'tests']
-  // This way, if we're given 'All.Passing' as a module name, we can also
-  // flip it into ['Passing', 'All'], and see if the first N elements line up.
-  return parsed.dir
-    .split(path.sep)
-    .concat([basename])
-    .filter(function (segment) {
-      return segment !== '';
-    })
-    .reverse();
-}
+function findTests(
+  testFilePaths /*: Array<string> */,
+  sourceDirs /*: Array<string> */,
+  isPackageProject /*: boolean */
+) /*: Promise<Array<{ moduleName: string, possiblyTests: Array<string> }>> */ {
+  return Promise.all(
+    testFilePaths.map((filePath) => {
+      const matchingSourceDirs = sourceDirs.filter((dir) =>
+        filePath.startsWith(`${dir}${path.sep}`)
+      );
+
+      // Tests must be in tests/ or in source-directories – otherwise they won’t
+      // compile. Elm won’t be able to find imports.
+      switch (matchingSourceDirs.length) {
+        case 0:
+          return Promise.reject(
+            Error(missingSourceDirectoryError(filePath, isPackageProject))
+          );
 
-function moduleFromTestName(testName) {
-  return testName.split('.').reverse();
-}
+        case 1:
+          // Keep going.
+          break;
 
-function filterMap(array, f) {
-  return array.map(f).filter(Boolean);
+        default:
+          // This shouldn’t be possible for package projects.
+          return Promise.reject(
+            new Error(
+              multipleSourceDirectoriesError(filePath, matchingSourceDirs)
+            )
+          );
+      }
```

Comment:
I wrote this out of necessity (previously we got module names from elmi-to-json. Now we need to figure them out ourselves), but it has a nice bonus effect: If a user runs `elm-test custom-test-dir/` and `custom-test-dir/` isn’t in `"source-directories"` they will now get a much better error message.

Reply:
Nice!